### DISCUSSION TITLE
Parameterset HM-CC-RT-DN Format - Temperatur

### DISCUSSION THREAD

**Discussion by morlie** [SukramJ/aiohomematic#1660] [Q&A] [answered]
Hi, ich nutze das Plugin "Homematic local" in Version 1.64.0. Ich weiß nicht ob es ein Bug ist oder ein falscher Aufruf. Es ist ein HM-CC-RT-DN, altes Homematic (bevor IP).

Wahrscheinlich irrelevant, Homematic läuft auf einer Debmatic installation. 

Ich kann die Temperatur nicht setzen, eine Idee woran es liegt?

Wenn ich Paramset nutze (Parametersatz schreiben) werden alle Parameter außer die Temperatur gesetzt. 
> ENDTIME_SATURDAY_1: 1440
TEMPERATURE_SATURDAY_1: 12
VALVE_OFFSET: 0

Format selber Parameter mit "Parametersatz lesen" sehen genau so aus.

> ENDTIME_SATURDAY_1: 1440
TEMPERATURE_SATURDAY_1: 11
VALVE_OFFSET: 0

Beispielaufruf

> params:
  domain: homematicip_local
  service: put_paramset
  service_data:
    paramset_key: MASTER
    rx_mode: BURST
    wait_for_callback: 40
    device_id: f1ebe8a2b7b58d0c4289d52cba65cc96
    paramset:
      TEMPERATURE_MONDAY_1: 12
      TEMPERATURE_TUESDAY_1: 12
      TEMPERATURE_WEDNESDAY_1: 12
      TEMPERATURE_THURSDAY_1: 12
      TEMPERATURE_FRIDAY_1: 12
      TEMPERATURE_SATURDAY_1: 12
      ENDTIME_SATURDAY_1: 1340
      TEMPERATURE_SUNDAY_1: 12
      VALVE_OFFSET: 0
  target: {}

Selbst wenn ich den Parametersatz auf "Get Paramset" identisch in "Put Paramset" nur mit geänderten Temperaturwerten poste wird die Temperatur nicht angepasst. Werte wie ENDTIME* oder VALVE_OFFSET werden übernommen.

> 2024-08-25 10:58:54.449 DEBUG (MainThread) [hahomematic.client] PUT_PARAMSET: REQ0844982, MASTER, {'ENDTIME_SATURDAY_1': 1440, 'TEMPERATURE_MONDAY_1': 12, 'TEMPERATURE_TUESDAY_1': 12, 'TEMPERATURE_WEDNESDAY_1': 12, 'TEMPERATURE_THURSDAY_1': 12, 'TEMPERATURE_FRIDAY_1': 12, 'TEMPERATURE_SATURDAY_1': 12, 'TEMPERATURE_SUNDAY_1': 12, 'VALVE_OFFSET': 0}
2024-08-25 10:58:54.449 DEBUG (MainThread) [hahomematic.client.xml_rpc] __ASYNC_REQUEST: ('putParamset', ('REQ0844982', 'MASTER', {'ENDTIME_SATURDAY_1': 1440, 'TEMPERATURE_MONDAY_1': 12, 'TEMPERATURE_TUESDAY_1': 12, 'TEMPERATURE_WEDNESDAY_1': 12, 'TEMPERATURE_THURSDAY_1': 12, 'TEMPERATURE_FRIDAY_1': 12, 'TEMPERATURE_SATURDAY_1': 12, 'TEMPERATURE_SUNDAY_1': 12, 'VALVE_OFFSET': 0}, 'BURST'))
2024-08-25 10:58:54.523 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 2.2e-05s ()
2024-08-25 10:58:54.523 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 1.4e-05s ()
2024-08-25 10:58:54.523 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 1e-05s ()
2024-08-25 10:58:54.523 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 8e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 8e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 7e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 6e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 6e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of _track_single_entity_state_change_or_timeout took 8e-06s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of wait_for_state_change_or_timeout took 0.00107s ()
2024-08-25 10:58:54.524 INFO (MainThread) [hahomematic.performance] Execution of put_paramset took 0.075105s (interface_id: Debmatic-BidCos-RF)

**Comment by SukramJ** [comment]
1. Homematic(IP) Local ist eine Integration und kein Plugin! Plugins gibt es bei HA nicht.
2. Was hast Du vor? Willst Du nur die Solltemperatur setzten? Ein bisschen Erklärung wäre sinnvoll.

**Reply by morlie** [reply]
Richtig, ich möchte nur die Solltemperaturen der Heizprofile im Thermostat unter den Beispielsparametern setzen.  Sollte kein großer Akt sein. Kompletter Parametrsatz:

> context: id: 01J64AB6PR1Q8JSNPEV3N09PZB parent_id: 01J64AB6PRWVR2YYEH8W8G4CPK user_id: null paramset: ADAPTIVE_REGULATION: 2 BACKLIGHT_ON_TIME: 10 BOOST_AFTER_WINDOW_OPEN: false BOOST_POSITION: 70 BOOST_TIME_PERIOD: 1 BURST_RX: true BUTTON_LOCK: false BUTTON_RESPONSE_WITHOUT_BACKLIGHT: false CYCLIC_INFO_MSG: true CYCLIC_INFO_MSG_DIS: 0 DAYLIGHT_SAVING_TIME: true DECALCIFICATION_TIME: 660 DECALCIFICATION_WEEKDAY: 0 DISPLAY_INFORMATION: 0 ENDTIME_FRIDAY_1: 1440 ENDTIME_FRIDAY_10: 1440 ENDTIME_FRIDAY_11: 1440 ENDTIME_FRIDAY_12: 1440 ENDTIME_FRIDAY_13: 1440 ENDTIME_FRIDAY_2: 960 ENDTIME_FRIDAY_3: 1440 ENDTIME_FRIDAY_4: 1440 ENDTIME_FRIDAY_5: 1440 ENDTIME_FRIDAY_6: 1440 ENDTIME_FRIDAY_7: 1440 ENDTIME_FRIDAY_8: 1440 ENDTIME_FRIDAY_9: 1440 ENDTIME_MONDAY_1: 1440 ENDTIME_MONDAY_10: 1440 ENDTIME_MONDAY_11: 1440 ENDTIME_MONDAY_12: 1440 ENDTIME_MONDAY_13: 1440 ENDTIME_MONDAY_2: 1200 ENDTIME_MONDAY_3: 1440 ENDTIME_MONDAY_4: 1320 ENDTIME_MONDAY_5: 1440 ENDTIME_MONDAY_6: 1440 ENDTIME_MONDAY_7: 1440 ENDTIME_MONDAY_8: 1440 ENDTIME_MONDAY_9: 1440 ENDTIME_SATURDAY_1: 1440 ENDTIME_SATURDAY_10: 1440 ENDTIME_SATURDAY_11: 1440 ENDTIME_SATURDAY_12: 1440 ENDTIME_SATURDAY_13: 1440 ENDTIME_SATURDAY_2: 1200 ENDTIME_SATURDAY_3: 1440 ENDTIME_SATURDAY_4: 1440 ENDTIME_SATURDAY_5: 1440 ENDTIME_SATURDAY_6: 1440 ENDTIME_SATURDAY_7: 1440 ENDTIME_SATURDAY_8: 1440 ENDTIME_SATURDAY_9: 1440 ENDTIME_SUNDAY_1: 1440 ENDTIME_SUNDAY_10: 1440 ENDTIME_SUNDAY_11: 1440 ENDTIME_SUNDAY_12: 1440 ENDTIME_SUNDAY_13: 1440 ENDTIME_SUNDAY_2: 1320 ENDTIME_SUNDAY_3: 1440 ENDTIME_SUNDAY_4: 1440 ENDTIME_SUNDAY_5: 1440 ENDTIME_SUNDAY_6: 1440 ENDTIME_SUNDAY_7: 1440 ENDTIME_SUNDAY_8: 1440 ENDTIME_SUNDAY_9: 1440 ENDTIME_THURSDAY_1: 1440 ENDTIME_THURSDAY_10: 1440 ENDTIME_THURSDAY_11: 1440 ENDTIME_THURSDAY_12: 1440 ENDTIME_THURSDAY_13: 1440 ENDTIME_THURSDAY_2: 1020 ENDTIME_THURSDAY_3: 1440 ENDTIME_THURSDAY_4: 1440 ENDTIME_THURSDAY_5: 1440 ENDTIME_THURSDAY_6: 1440 ENDTIME_THURSDAY_7: 1440 ENDTIME_THURSDAY_8: 1440 ENDTIME_THURSDAY_9: 1440 ENDTIME_TUESDAY_1: 1440 ENDTIME_TUESDAY_10: 1440 ENDTIME_TUESDAY_11: 1440 ENDTIME_TUESDAY_12: 1440 ENDTIME_TUESDAY_13: 1440 ENDTIME_TUESDAY_2: 1200 ENDTIME_TUESDAY_3: 1440 ENDTIME_TUESDAY_4: 1320 ENDTIME_TUESDAY_5: 1440 ENDTIME_TUESDAY_6: 1440 ENDTIME_TUESDAY_7: 1440 ENDTIME_TUESDAY_8: 1440 ENDTIME_TUESDAY_9: 1440 ENDTIME_WEDNESDAY_1: 1440 ENDTIME_WEDNESDAY_10: 1440 ENDTIME_WEDNESDAY_11: 1440 ENDTIME_WEDNESDAY_12: 1440 ENDTIME_WEDNESDAY_13: 1440 ENDTIME_WEDNESDAY_2: 1020 ENDTIME_WEDNESDAY_3: 1440 ENDTIME_WEDNESDAY_4: 1320 ENDTIME_WEDNESDAY_5: 1440 ENDTIME_WEDNESDAY_6: 1440 ENDTIME_WEDNESDAY_7: 1440 ENDTIME_WEDNESDAY_8: 1440 ENDTIME_WEDNESDAY_9: 1440 GLOBAL_BUTTON_LOCK: false I_VALUE_EXTERN: 15 I_VALUE_INTERN: 15 LOCAL_RESET_DISABLE: false LOW_BAT_LIMIT: 2.1 MANU_MODE_PRIORITIZATION: 1 MIN_MAX_VALUE_NOT_RELEVANT_FOR_MANU_MODE: false MODUS_BUTTON_LOCK: false PARTY_MODE_PRIORITIZATION: 1 P_START_VALUE_EXTERN: 30 P_START_VALUE_INTERN: 30 P_VALUE_EXTERN: 30 P_VALUE_INTERN: 30 SHOW_WEEKDAY: false TEMPERATUREFALL_MODUS: 4 TEMPERATUREFALL_VALUE: 1.4 TEMPERATUREFALL_WINDOW_OPEN: 12 TEMPERATUREFALL_WINDOW_OPEN_TIME_PERIOD: 15 TEMPERATURE_COMFORT: 15 TEMPERATURE_FRIDAY_1: 8 TEMPERATURE_FRIDAY_10: 17 TEMPERATURE_FRIDAY_11: 17 TEMPERATURE_FRIDAY_12: 17 TEMPERATURE_FRIDAY_13: 17 TEMPERATURE_FRIDAY_2: 21.5 TEMPERATURE_FRIDAY_3: 20.5 TEMPERATURE_FRIDAY_4: 18 TEMPERATURE_FRIDAY_5: 17 TEMPERATURE_FRIDAY_6: 17 TEMPERATURE_FRIDAY_7: 17 TEMPERATURE_FRIDAY_8: 17 TEMPERATURE_FRIDAY_9: 17 TEMPERATURE_LOWERING: 10 TEMPERATURE_MAXIMUM: 30.5 TEMPERATURE_MINIMUM: 4.5 TEMPERATURE_MONDAY_1: 8 TEMPERATURE_MONDAY_10: 17 TEMPERATURE_MONDAY_11: 17 TEMPERATURE_MONDAY_12: 17 TEMPERATURE_MONDAY_13: 17 TEMPERATURE_MONDAY_2: 20 TEMPERATURE_MONDAY_3: 17 TEMPERATURE_MONDAY_4: 21 TEMPERATURE_MONDAY_5: 17 TEMPERATURE_MONDAY_6: 17 TEMPERATURE_MONDAY_7: 17 TEMPERATURE_MONDAY_8: 17 TEMPERATURE_MONDAY_9: 17 TEMPERATURE_OFFSET: 7 TEMPERATURE_SATURDAY_1: 8 TEMPERATURE_SATURDAY_10: 17 TEMPERATURE_SATURDAY_11: 17 TEMPERATURE_SATURDAY_12: 17 TEMPERATURE_SATURDAY_13: 17 TEMPERATURE_SATURDAY_2: 21 TEMPERATURE_SATURDAY_3: 17 TEMPERATURE_SATURDAY_4: 17 TEMPERATURE_SATURDAY_5: 17 TEMPERATURE_SATURDAY_6: 17 TEMPERATURE_SATURDAY_7: 17 TEMPERATURE_SATURDAY_8: 17 TEMPERATURE_SATURDAY_9: 17 TEMPERATURE_SUNDAY_1: 8 TEMPERATURE_SUNDAY_10: 17 TEMPERATURE_SUNDAY_11: 17 TEMPERATURE_SUNDAY_12: 17 TEMPERATURE_SUNDAY_13: 17 TEMPERATURE_SUNDAY_2: 17 TEMPERATURE_SUNDAY_3: 18 TEMPERATURE_SUNDAY_4: 17 TEMPERATURE_SUNDAY_5: 17 TEMPERATURE_SUNDAY_6: 17 TEMPERATURE_SUNDAY_7: 17 TEMPERATURE_SUNDAY_8: 17 TEMPERATURE_SUNDAY_9: 17 TEMPERATURE_THURSDAY_1: 8 TEMPERATURE_THURSDAY_10: 17 TEMPERATURE_THURSDAY_11: 17 TEMPERATURE_THURSDAY_12: 17 TEMPERATURE_THURSDAY_13: 17 TEMPERATURE_THURSDAY_2: 19.5 TEMPERATURE_THURSDAY_3: 20.5 TEMPERATURE_THURSDAY_4: 18 TEMPERATURE_THURSDAY_5: 17 TEMPERATURE_THURSDAY_6: 17 TEMPERATURE_THURSDAY_7: 17 TEMPERATURE_THURSDAY_8: 17 TEMPERATURE_THURSDAY_9: 17 TEMPERATURE_TUESDAY_1: 8 TEMPERATURE_TUESDAY_10: 17 TEMPERATURE_TUESDAY_11: 17 TEMPERATURE_TUESDAY_12: 17 TEMPERATURE_TUESDAY_13: 17 TEMPERATURE_TUESDAY_2: 20 TEMPERATURE_TUESDAY_3: 17 TEMPERATURE_TUESDAY_4: 21 TEMPERATURE_TUESDAY_5: 17 TEMPERATURE_TUESDAY_6: 17 TEMPERATURE_TUESDAY_7: 17 TEMPERATURE_TUESDAY_8: 17 TEMPERATURE_TUESDAY_9: 17 TEMPERATURE_WEDNESDAY_1: 8 TEMPERATURE_WEDNESDAY_10: 17 TEMPERATURE_WEDNESDAY_11: 17 TEMPERATURE_WEDNESDAY_12: 17 TEMPERATURE_WEDNESDAY_13: 17 TEMPERATURE_WEDNESDAY_2: 17 TEMPERATURE_WEDNESDAY_3: 19 TEMPERATURE_WEDNESDAY_4: 21 TEMPERATURE_WEDNESDAY_5: 17 TEMPERATURE_WEDNESDAY_6: 17 TEMPERATURE_WEDNESDAY_7: 17 TEMPERATURE_WEDNESDAY_8: 17 TEMPERATURE_WEDNESDAY_9: 17 VALVE_ERROR_RUN_POSITION: 15 VALVE_MAXIMUM_POSITION: 30 VALVE_OFFSET: 10

**Reply by SukramJ** [reply]
Lass die folgenden Parameter weg:
```
rx_mode: BURST
wait_for_callback: 40
```

Solche Fragen bitte im Homematic Forum stellen.
Da im Log kein Fehler auftaucht war die CCU wohl mit dem Paramset zufrieden, auch wenn es wohl nicht verarbeitet wurde.
Ob die übertragenen Werte für eine CCU semantisch korrekt sind kann ich die nicht sagen.

**Reply by SukramJ** [reply]
Bei meinem HmIP-BWTH wird die folgender Befehl verarbeitet und der Wert verändert:

```
action: homematicip_local.put_paramset
data:
  device_id: ea82e74e996057270f40858f48f0d425
  channel: 1
  paramset_key: MASTER
  paramset:
    P2_TEMPERATURE_MONDAY_1: 17
```

**Reply by Baxxy13** [reply]
Des Rätsels Lösung ist relativ simpel.
`user.err rfd: HSSParameter::SetValue() 13 exception type error`

HM-Geräte wollen da (im Gegensatz zu IP) zwingend float haben, also...
`P2_TEMPERATURE_MONDAY_1: 17` geht nicht
`P2_TEMPERATURE_MONDAY_1: 17.0` geht nicht
`P2_TEMPERATURE_MONDAY_1: 17.5` geht

Edit: 
Der Umweg über die virt. HM-Heizgruppen klappt auch nicht. Die Gruppen verdauen zwar den Integerwert, können ihn aber nicht an die Thermostate übermitteln.

Grundsätzlich sind die Datentypen der Temperaturen in den Profilen mit double (float) angegeben, egal ob HM/IP.

**Reply by morlie** [reply]
> Grundsätzlich sind die Datentypen der Temperaturen in den Profilen mit double (float) angegeben, egal ob HM/IP.

Das teste ich später. Hört sich schlüssig an, danke schonmal dafür. 

Gibt es eine Möglichkeit den Datentyp zu erzwingen?

**Reply by Baxxy13** [reply]
Da ist @SukramJ gefragt. 
~~Bei `homematicip_local.set_device_value` kann man ja explizit den Datentyp mitgeben, wäre hier vielleicht ne Option.~~
Oder float muss erzwungen werden wenn ich float eingebe, also bei z.B. `12.0`.
Wird schon werden. 😉 

Edit:
Datentyp mitgeben ist doch eher unsinnig da ein Parametersatz verschiedene Datentypen enthalten kann.

**Reply by morlie** [reply]
Muss mal 12.01 testen was  passiert

**Reply by Baxxy13** [reply]
Nicht schön, funktioniert aber als Workaround.

**Reply by morlie** [reply]
> Muss mal 12.01 testen was passiert

Habe das als Workaround implementiert und funktioniert. Durch deinen Input hast du mir geholfen, danke. Nicht sauber aber damit kann ich (erstmal) leben. Empfehlenswert wäre hier sicher als Wert ein float mitzugeben, also 12.0 und das beizubehalten. Damit es konsistent ist müsste auch bei "get paramset" ein float zurückgeliefert werden. Da es bis jetzt noch keinem aufgefallen ist bin ich vermutlich der erste der das überhaupt für diesen Zweck nutzt. Gibt sicher wichtigere Dinge zu ändern oder einzubauen.

**Reply by SukramJ** [reply, marked as answer]
Die nächste Version wird eine Validierung/Konvertierung gegen die Gerätebeschreibung durchführen, so dass der Workaround dann nicht mehr notwendig sein sollte.

**Reply by morlie** [reply]
> Die nächste Version wird eine Validierung/Konvertierung gegen die Gerätebeschreibung durchführen, so dass der Workaround dann nicht mehr notwendig sein sollte.

Vielen Dank!

**Reply by SukramJ** [reply]
Die Beta 1.65.0b10 steht jetzt zur Verfügung. Bitte testen.
Fehler stehen ggf. im Log

**Reply by morlie** [reply]
> Die Beta 1.65.0b10 steht jetzt zur Verfügung. Bitte testen. Fehler stehen ggf. im Log

Getestet, funktioniert
